allspice/utils/bom_generation.py

```diff
@@ -414,3 +452,132 @@ def _group_entries(
+    raise ValueError(
+        f"Variant {variant} not found in PrjPcb file.\n"
+        f"Available variants: {', '.join(available_variants)}"
+    )
+
+
+def _apply_variations(
+    components: list[dict[str, str | None]],
+    variant_details: configparser.SectionProxy,
+    logger: Logger,
+) -> list[dict[str, str | None]]:
+    """
+    Apply the variations of a specific variant to the components. This should be
+    done before the components are mapped to columns or grouped.
+
+    :param components: The components to apply the variations to.
+    :param variant_details: The section of the config file dealing with a
+        specific variant.
+
+    :returns: The components with the variations applied.
+    """
+
+    # Each item in the list is a pairing of (unique_id, designator), as both are
+    # required to identify a component.
+    components_to_remove: list[tuple[str, str]] = []
+    # When patching components, the ParamVariation doesn't have the unique ID,
+    # only a designator. However, ParamVariations follow the Variation entry, so
+    # if we note down the last unique id we saw for a designator when going
+    # through the variations, we can use that unique id when handling a param
+    # variation. This dict holds that information.
+    patch_component_unique_id: dict[str, str] = {}
+    # The keys are the same as above, and the values are a key-value of the
+    # parameter to patch and the value to patch it to.
+    components_to_patch: dict[tuple[str, str], tuple[str, str]] = {}
+
+    for key, value in variant_details.items():
+        # Note that this is in lowercase, as configparser stores all keys in
+        # lowercase.
+        if re.match(r"variation[\d+]", key):
+            variation_details = dict(details.split("=", 1) for details in value.split("|"))
+            try:
+                designator = variation_details["Designator"]
+                # The unique ID field is a "path" separated by backslashes, and
+                # the the unique id we want is the last entry in that path.
+                unique_id = variation_details["UniqueId"].split("\\")[-1]
+                kind = VariationKind(int(variation_details["Kind"]))
+                if kind == VariationKind.NOT_FITTED:
+                    components_to_remove.append((unique_id, designator))
+                else:
+                    patch_component_unique_id[designator] = unique_id
+            except KeyError:
+                logger.warn(
+                    "Designator, UniqueId, or Kind not found in details of variation "
+                    f"{variation_details}"
+                )
+                continue
+            except ValueError:
+                logger.warn(
+                    f"Kind {variation_details['Kind']} of variation {variation_details} must be "
+                    "either 0, 1 or 2."
+                )
+                continue
+        elif re.match(r"paramvariation[\d]+", key):
```

Comment:
We should pull out regex compilation. It's much slower than matching and will dominate the runtime of this function.

Reply:
According to they python docs, regexes are compiled and LRU cached, so there shouldn't be a performance difference here. Whether or not we `re.compile`, the code path will be ~~the same.

Reply:
Okay. It's fine if you don't want to change it. I, personally, wouldn't rely on that.